### PR TITLE
Use area-weighting when converting wind stress from centers to edges

### PR DESCRIPTION
**Pull request by sbrus89**
This PR addresses issue #215. When computing the wind stress on an edge from the adjacent cell centered values, an area-weighted average (based on edge kite area) is used instead of the mean value.

### CODE REVIEW

File: src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F [hunk 1]
```diff
@@ -198,11 +212,36 @@ subroutine ocn_surface_bulk_forcing_vel(meshPool, forcingPool, surfaceStress, su
...
+        z3 = zVertex(vertex2)
+
+        do iCell = 1,2
+          cell = cellsOnEdge(iCell, iEdge)
+          
+          x1 = xCell(cell)
+          y1 = yCell(cell)
+          z1 = zCell(cell)
+
+          areaWeights(iCell) = 0.5_RKIND*(sqrt(((y2-y1)*(z3-z1)-(y3-y1)*(z2-z1))**2+ &
```
Comment (xylar) [inline src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F]: @sbrus89, the area of the "kite" between edges is `dcEdge*dvEdge`.  I think you could use this (or half of this) as your area and save a *lot* of computation.

**Comment by sbrus89** [comment]
@pwolfram and I discussed this a bit and decided to open it up to anyone else who has feedback. The current approach requires a kiteAreasOnEdge array for the area-weights, which is currently not computed in MPAS (as far as I know). Right now I'm recomputing this each time and it should eventually be pre-computed. The question is where pre-computing this would fit best in the code.

We also discussed using the cellArea values as weights, which is a simpler approach.

Here's a diagram of the implementation at present:

wind_stress_averaging

**Comment by sbrus89** [comment]
I also realize this won't be BFB in E3SM, so I understand any hesitation to adopt this.

**Comment by xylar** [comment]
Believe edges are required to be half way between cell centers, so the red triangles you have drawn above are required to be identical, each with an area 0.5*dcEdge*dvEdge.

**Comment by xylar** [comment]
I could also be wrong, since I think Jigsaw relaxes some of these requirements.  @dengwirda, do you have thoughts?

**Comment by xylar** [comment]
Here's the mesh spec for reference: https://mpas-dev.github.io/files/documents/MPAS-MeshSpec.pdf

File: src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F (hunk 1 above)
Comment (pwolfram) [reply src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F]: Thanks @xylar, good call.

**Comment by sbrus89** [comment]
If that's the case, this might not even be needed.

**Comment by xylar** [comment]
> If that's the case, this might not even be needed.

Yes, actually, I think you're right.  The areas are supposed to be the same from either side of an edge so no weighting should be needed.  Again, Jigsaw might relax this a bit but I doubt it's a major source of error even so.  And this would almost certainly not be the only place in the code that makes this assumption.

File: src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F (hunk 1 above)
Comment (pwolfram) [reply src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F]: But, dcEdge is between cell centers.  So, this implicitly is averaging for the area computation, correct?

**Comment by pwolfram** [comment]
The areas of the red triangles cannot be the same, the height of them from the edge to cell mid-points differs.  So, dcEdge*dvEdge is already doing some averaging and in that case the 0.5 factor emerges.

File: src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F (hunk 1 above)
Comment (pwolfram) [reply src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F]: I don't think we can simplify like this via `dcEdge*dvEdge`; we need the full triangle contained in the hexagon to be computed.

**Comment by xylar** [comment]
@pwolfram, for the Voronoi mesh, the segment between vertices is defined to be perpendicular to the segment between cell centers.  The edge is exactly half way between cell centers.  That means that the two red triangles are of equal area: each is 0.5*dcEdge*dvEdge.

**Comment by pwolfram** [comment]
@xylar, I don't believe this is true but is except for constant resolution meshes.  We can make the cell on one side bigger, which moves the cell center, but the edge remains fixed.  This means that the distance from the edge mid-point to the cell center on one side has now increased or decreased relative to the other side that remained fixed.

> Believe edges are required to be half way between cell centers, so the red triangles you have drawn above are required to be identical, each with an area 0.5dcEdgedvEdge.

**Comment by xylar** [comment]
Note: we had a call about this and @pwolfram and @sbrus89 will look into this further with regard to their coastal meshes.  @pwolfram is still not completely convinced so further discussion may be needed.

**Comment by pwolfram** [comment]
Thanks @xylar, you are correct and thank you for clearing this up-- we conflated triangles with hexagons based on coming from triangular meshes originally.  You can see this property is an issue on the triangular mesh, but not on the hexagonal mesh (see below for illustration).

Screen Shot 2019-05-03 at 9 33 35 AM

The outstanding question for @dengwirda is whether this property will be relaxed under a Hodge Optimized Triangulation as he has been exploring.

File: src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F (hunk 1 above)
Comment (pwolfram) [reply src/core_ocean/shared/mpas_ocn_surface_bulk_forcing.F]: Thanks @xylar!

**Comment by xylar** [comment]
I suggest closing this at least for now.

**Comment by sbrus89** [comment]
Agreed

**Comment by dengwirda** [comment]
Okay, I think there's a few things worth considering here!

Firstly, I believe that @xylar is correct -- by definition each Voronoi edge is guaranteed to bisect the Delaunay edge it's paired with, so the area of the triangles you mention should be the same.

The opposite is not true though -- Delaunay edges don't necessarily bisect Voronoi edges, except when the mesh consists of all regular hexagons. I think this is essentially what @pwolfram is getting at, and can be seen in the figure.

A particularly bad case can occur when obtuse triangles are involved -- in these cases the Voronoi and Delaunay edges actually don't intersect at all (though the projection of the Voronoi edge would still hit the Delaunay midpoint).

Generally speaking `JIGSAW` is reasonably good at generating meshes without any obtuse triangles, even in cases where the resolution is changing, so this is hopefully not too much of an issue in practice, though probably worth keeping in mind.

**

The weighted case (i.e. Hodge optimised) is different though -- here constraints are relaxed, and neither edge is guaranteed to bisect the other (they are still orthogonal though).

I won't go on too much about this, but the general idea is that by distributing the offset between both the triangle and cell edges, it seems to be possible to generate meshes that are more regular overall. For example, it's possible to have obtuse triangles in weighted meshes and for all of the cell and triangle edges to intersect properly.

`MPAS` is currently only using unweighted meshes (i.e. Voronoi / Delaunay grids), so dealing with the weighted geometry isn't currently a live issue. 

To support Hodge-type meshes in the future though, I believe the full set of calculations (as per @sbrus89) would be needed.

**Comment by sbrus89** [comment]
Thanks @dengwirda! We can leave this on the back-burner and keep it in mind for future reference.

**Pull request closed without merging**